### ISSUE TITLE
Relative links without a leading forward slash are broken when previewing content changes

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5023]
## Steps to reproduce
1. On a vanilla installation of Backdrop, start creating a post.
2. Add a title.
3. In the body field, switch the text format to "Raw HTML".
4. Paste the following in the body field: `<p>This is a <a href="posts/your-first-post">link to the first post</a> of this site.</p>`.
5. Scroll to the bottom of the page and click "Preview".
6. In the preview page that loads, open the link in another browser tab, to test it -> access denied 👎🏼 
7. Back in the preview window, hover over the link -> the URL is `https://your.test.site/node/preview/page/posts/your-first-post` 👎🏼 (the `node/preview/page/` bit should not be added).
8. Click "Back to content editing".
9. Edit the content of the body field, so that the href in the link includes a leading forward slash (`/`). The code should no be like so: `<p>This is a <a href="/posts/your-first-post">link to the first post</a> of this site.</p>`
10. Scroll to the bottom of the page and click "Preview".
11. In the preview page that loads, hover over the link -> the URL is `https://your.test.site/posts/your-first-post` 👍🏼 
12. Test the link by opening it in a new tab/window -> the content of the "Your first post" node loads without any error 👍🏼 

---

## Original report

I noticed this when I was trying to update https://backdropcms.org/roadmap#2.0 by linking it to https://backdropcms.org/news/backdrop-cms-version-2-will-not-be-released-before-january-2025

I added this:

`<a href="news/backdrop-cms-version-2-will-not-be-released-before-january-2025"> ... </a>`

...which when previewing the node results in this URL when rendering the link:

`https://backdropcms.org/node/preview/book/news/backdrop-cms-version-2-will-not-be-released-before-january-2025`

Raising this issue, to see if we can make link rendering "smart" enough to detect that this is a preview page, and in that case NOT add the `node/preview/book/` bit in the links.

**Comment by indigoxela** [comment]
Did you add the link via the editor widget, or was it markup?

It's not that we don't have enough issues for link handling. :smile:

But honestly, I don't think you should add links like that (without base path). In case a node alias changes to use a "path", this link would break anyway.

**Comment by klonos** [comment]
> Did you add the link via the editor widget, or was it markup?

You're right @indigoxela ...I've done some more testing though, and if you add a link to an internal page via the CKEditor button, you end up with markup like `<p>test <a href="/about">link</a> here</p>`

**Comment by klonos** [comment]
...perhaps it's specific to how filters are configured in b.org(?)

**Comment by klonos** [comment]
> ...perhaps it's specific to how filters are configured in b.org(?)

The text format in b.org doesn't use any text editor. That seems to be the cause.

**Comment by indigoxela** [comment]
So this is nothing, core could solve, right? It's not a bug IMO.

**Comment by klonos** [comment]
If the functionality that converts things to links and handles them properly for normal display as well as for preview is CKEditor, then yes, I suppose so. If that functionality is supposed to be handled by text filters, independently of which editor (or no editor), then we should look into it. Right?

**Comment by indigoxela** [comment]
> If that functionality is supposed to be handled by text filters

Hm, is that the case? I read your issue description like "you manually inserted that link into content". Maybe I missed something, but if you manually add a relative link as markup, which filter could fix a possible mistake?

**Comment by philsward** [comment]
If I understand correctly, this goes back to the roots of Drupal.

If a link does not have the preceding / then it takes it off of the "relative" path. If you add the / it goes off the "absolute" path. (My terminology might be a bit off, sorry, hope you get the point)

I ran into this in early Drupal days when I would use a sub-domain for testing and the path had additional sub-paths all of my links would be goofy. (foo/bar/ vs foo-bar/)

I learned quick to ALWAYS precede the link with a /. Never had a problem since.

So, is it a bug? I don't think so, however I do think it might be a good idea for link filters to always assume the link needs the preceding /.

**Comment by klonos** [comment]
> If a link does not have the preceding / then it takes it off of the "relative" path. If you add the / it goes off the "absolute" path.

This is actually exactly what happened 👍🏼 ...I did not add a leading `/` the time I had the issue.

> So, is it a bug? I don't think so, however I do think it might be a good idea for link filters to always assume the link needs the preceding /.

That's why I'd like us to look into this a bit more at some point.

**Comment by indigoxela** [comment]
Maybe this discussion should go on in the backdropcms.org issue queue?

The link has been added manually, so it isn't created by a filter (like the url filter) when the content gets rendered, but it exists in markup.

Fixing things that people added manually (as markup) currently isn't possible AFAIK. If the link had been added via the editor dialog, then the dialog would have had a chance to fix things (note, we already have an [Issue for that dialog](https://github.com/backdrop/backdrop-issues/issues/4630) which is pending testing and review).

But the situation here is slightly different. To my understanding the request here is to add a "link fixer filter", which would require a http request for every link added, and to correct (?) them if they're not working - that very complex IMO.

**Comment by klonos** [comment]
>Maybe this discussion should go on in the backdropcms.org issue queue?

@indigoxela I've updated the issue summary, to include steps to reproduce this issue in a vanilla Backdrop installation.

> Fixing things that people added manually (as markup) currently isn't possible AFAIK.

The link was added using the "Raw HTML" text format that ships with Backdrop OOTB. That format includes the "Correct faulty and chopped off HTML" filter, which I think is where we should be fixing this. This is basically the issue as @philsward reported having faced before in Drupal:

> If a link does not have the preceding / then it takes it off of the "relative" path. If you add the / it goes off the "absolute" path. ... I learned quick to ALWAYS precede the link with a /. Never had a problem since.

What I'm suggesting is to make it so that this leading `/` is added automatically if missing (from paths - not from proper URLs that include protocol etc.). This would save people that aren't long-time Drupal users from having to know that the missing `/` is an issue -> UX win.

**Comment by indigoxela** [comment]
> The link was added using the "Raw HTML" text format that ships with Backdrop OOTB. That format includes the "Correct faulty and chopped off HTML" filter, which I think is where we should be fixing this.

I disagree, the markup is syntactically correct, so this filter isn't responsible. Its purpose is to fix tags that, for example, aren't properly closed.

Fixing links that *are* correct markup, just don't point to where they should to, needs a totally new filter.
And it's very likely necessary to turn off that filter individually - fiddling with user-entered stuff when rendered can lead to way more WTF moments. They _might_ have entered a relative url by intention and get baffled that it "got mangled". :wink:

**Comment by klonos** [comment]
Interesting: https://www.daymuse.com/blogs/drupal-broken-internal-link-path-module-tutorial

> If you were to simply add the href of "page-name-2" to a link from a page at this URL:
> 
> `http(s)://subdomain.domain.extension/topic/page-name`
>
> The href would be relative to the current location and so the link will actually go to:
> 
> `http(s)://subdomain.domain.extension/topic/page-name​-2`
>
> The link destination will retain the depth, and target a sibling page ("page-name-2" within /topic/).
> 
> You could also link to a page on the current domain from the top level by using a leading forward slash. If your link href was set to "/page-name-2" in the above example, with a leading forward slash, your destination would become: 
> 
> `http(s)://subdomain.domain.extension/page-name​-2`
> The forward slash causes the hierarchy to collapse, leading from the top level domain: "/topic" is dropped.

Seems that there are valid use cases for relative paths to be both with and without a `/`, and that they serve a different purpose 🤔 ...I need to do some further testing, and see if we should close this as "works as designed", or do some special-casing with the content preview mode.

**Comment by philsward** [comment]
@klonos yeah, that's kind of what I found in my early days as well. It just works goofy...

The only fix I found to "always" make it work is put the preceding / with the full path (/topic/foo/bar)

Drupal is supposed to be smart about the links, but it caused me weird issues and  I doubt the "correct procedure" for the different scenarios are documented anywhere. (Never looked)

I have a feeling it is the way it is for a reason (so not a bug like @indigoxela points out), but I agree 100% that it isn't good UX.

**Comment by philsward** [comment]
Ah, I think the reason for both might be this use case:

Content type with alias of .com/foo
Page at .com/topic/foo/bar

and 

Content type with alias of .com/
Page at: .com/bar

If creating a page of the content type foo, and the link needs to point to .com/bar, you MUST have /bar

Otherwise without the /, the link gets translated to .com/topic/foo/bar

The problem with this I found is that if a link path aliases changes, it can cause problems both ways. In the case of using /topic/bar as the link and the topic is renamed to discussion, all links using /topic/*/* are broken.

If they do not use the preceding / and instead use bar as the link path, the alias path is updated.

I can't remember exactly what I did but when messing with moving the domain to a sandbox environment, links that did not precede with a / would break.

Maybe it was a sub-domain (doesn't really make sense) or maybe it was because I added the site into a folder of the domain (.com/sandbox/*). and didn't update the docroot. Honestly can't remember it's been so long.

**Comment by klonos** [comment]
> I don't feel like I'm being much help here, but I'm trying 😉

You've already helped many comments ago @philsward 👍🏼
